packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -459,21 +468,55 @@ export class Pipeline extends PipelineBase {
 
     // check whether we already have a bucket in that region,
     // either passed from the outside or previously created
+    const crossRegionSupport = this.obtainCrossRegionSupportFor(action, actionEnv.region);
+
+    // the stack containing the replication bucket must be deployed before the pipeline
+    Stack.of(this).addDependency(crossRegionSupport.stack);
+    // The Pipeline role must be able to replicate to that bucket
+    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+
+    return {
+      artifactBucket: crossRegionSupport.replicationBucket,
+      region: actionEnv.region,
+    };
+  }
+
+  /**
+   * Get or create the cross-region support construct for the given action
+   */
+  private obtainCrossRegionSupportFor(action: IAction, actionRegion: string) {
     let crossRegionSupport = this._crossRegionSupport[actionRegion];
     if (!crossRegionSupport) {
       // we need to create scaffolding resources for this region
+      const otherStack = this.crossRegionSupportScope(action);
       crossRegionSupport = this.createSupportResourcesForRegion(otherStack, actionRegion);
       this._crossRegionSupport[actionRegion] = crossRegionSupport;
     }
+    return crossRegionSupport;
+  }
 
-    // the stack containing the replication bucket must be deployed before the pipeline
-    pipelineStack.addDependency(crossRegionSupport.stack);
-    crossRegionSupport.replicationBucket.grantReadWrite(this.role);
+  /**
+   * Return the correct scope to create the cross-region construct in
+   *
+   * As an optimization, we reuse the action resource's stack if the stack has
+   * the same environment as the action.
+   *
+   * (Always true for owned action resources but not necessarily true
+   * for imported ones).
+   */
+  private crossRegionSupportScope(action: IAction): Stack | undefined {
```

Comment:
How about calling this `determineCrossRegionSupportScopeForAction`? Or similar. `crossRegionSupportScope` is a little too vague to my liking.